File: 678-valid-parenthesis-string/valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {
        int minOpen=0;
        int maxOpen=0;
        int n=s.length();
        for(int i=0; i<n ;i++){
            char ch=s.charAt(i); 
            if(ch=='('){
                minOpen+=1;
                maxOpen+=1;
                }

            else if(ch==')'){
                minOpen-=1;
                maxOpen-=1;
            } 
            else {
                minOpen-=1;
                maxOpen+=1;
            }  
            if(maxOpen<0){
                return false;
            }
            minOpen=Math.max(minOpen,0);
        }
        return minOpen==0;
    }
}

File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        openStack=[]
        starStack=[]

        for i,ch in enumerate(s):
            if ch=='(':
                openStack.append(i)
            elif ch=='*':
                starStack.append(i)
            else:
                if openStack:
                    openStack.pop()
                elif starStack:
                    starStack.pop()
                else:
                    return False

        while openStack and starStack:
            if openStack[-1]<starStack[-1]:
                openStack.pop()
                starStack.pop()
            else:
                return False
        
        return not openStack